### ISSUE TITLE
IPS patches are not working for KOF96 and other issues

### ISSUE THREAD

**Issue by dezraj** [finalburnneo/FBNeo#1454]
I'm not sure if it's helpful to report three issues in one bug report only, but here they are:

1) It looks like IPS patches are not working for KOF96. I'm using the following patches:

2-flash.dat
2-practi.dat
3-bossF.dat
8-time.dat

They are taken from these packs:
https://github.com/taoenwen/FBNeo_IPS
http://www.emu-france.com/emulateurs/6-arcades/89-final-burn/3794-ips-collection-v4-3-4-4-test/
https://www.youtube.com/watch?v=JewKnkptZe8 (See video description)

I have tried both 32 and 64 bit versions and none worked. The patches for KOF97 worked fine. The KOF96 patches also work fine in Raine emulator, as you see here:
![image](https://user-images.githubusercontent.com/27839354/235286283-76dfc138-01da-4490-8db5-43437cf3517d.png)
![image](https://user-images.githubusercontent.com/27839354/235286294-f14aa6a3-0a1d-4920-bdf7-8e4ab5a4a0d3.png)

2) The DirectX9 Alt is not working in the 64 bit version. I get the following message:
![image](https://user-images.githubusercontent.com/27839354/235286374-5898ea43-67f2-4c44-87e0-5b7957d47d41.png)

But I do have those DirectX runtimes installed, otherwise they wouldn't work in the 32 bit version.

3) If I select the SoftFX blitter and hq4x filter (32 bit only), I get the following result:
![image](https://user-images.githubusercontent.com/27839354/235286625-b5f09742-1172-4e38-820d-9f740328c49f.png)

If I still have the SoftFX blitter selected and set the 4xBR Rounded (16 bit only) filter there will be huge slowdown in the game. This won't happen if I'm in the Enhanced or DirectX9 Alt blitter (but the latter only works in 32 bit version of the program).

If I select SoftFX and Double pixels (Software) and try to enter full-screen, I get the following:
![image](https://user-images.githubusercontent.com/27839354/235287154-a3cc3bd6-75e0-4be3-adeb-5a11a3e69f81.png)

So basically, all blitters considered, the ones that work better here are the Enhanced and the Experimental. The DirectX9 Alt only works in the 32 bit version of the emulator as far as I've noticed. Perhaps the Enchanced blitter could be the default one then?

I hope this report helps improve the program somehow. Thank you so much in advance for your attention.

**Comment by taoenwen** [comment]
@dezraj Sorry, I didn't find any of these ips patch files in my repository, can you verify?

**Comment by dinkc64** [comment]
taoenwen:
it's in this big file: https://mega.nz/file/4IxwlS4J#waX-TZbX7eA1ChMpYQxc3d-7uebf5KeRetEAAwCyck4

here's just support/ips/kof96 subdir zip:
[kof96.zip](https://github.com/finalburnneo/FBNeo/files/11360669/kof96.zip)

**Comment by taoenwen** [comment]
@dinkc64 Thanks
@dezraj 
The current rom label for kof96's p1 in fbneo is: 214-p1.p1
The rom label of the p1 corresponding to the DAT in the ips patch you mentioned does not work is: 214-p1.bin
You need to change 214-p1.bin in the DAT file to 214-p1.p1

**Comment by dinkc64** [comment]
note2: I think problem is filename difference between datfile and d_neogeo.  f.ex: 214-p1.bin in .dat, 214-p1.p1 in d_neogeo.

about changing dat, lets wait.  tomorrow I will add crc32 check to IpsApplyPatches()  (unless you do it before me) :)  We should have had crc32 check to begin with.. :)

best regards,
- dink

**Comment by dezraj** [comment]
Hello @taoenwen and @dinkc64 ! I'm sorry for the late reply. Yes, I found those patches in this "Chinese" pack in this Youtube tutorial video for FBA. They have interesting features for K.O.F. games such as this one for KOF96 which removes the flash effect in the game. This feature was later added by SNK itself to KOF98 and newer games due to seizure complaints from the public. I don't suffer from seizure but they also hurt my vision a lot especially when I am in a dimmed room. If you could add them to your pack it would be a nice addition I guess.

@dinkc64 : Do you think something can be done to the issues I reported with the blitters and filters?

Anyway, thank you so much guys for your help and work.

**Comment by taoenwen** [comment]
@dezraj Here's the finished fix for you to try, and good luck with everything!
[kof96.zip](https://github.com/finalburnneo/FBNeo/files/11361284/kof96.zip)

**Comment by dinkc64** [comment]
The blitters are a mixed bag, each one runs better on some systems than others, that's all I can really say or do :)

**Comment by dinkc64** [comment]
Added a feature to the IPS manager to check the crc32, this will allow the older patches to work without modification, where the filename of the romfile has changed has changed.

**Comment by dezraj** [comment]
Thank you @taoenwen for the fixed IPS patches. They now work in the current release (the one I already had unzipped here). But I will update to the latest FBNeo version also.

@dinkc64 Would something be broken in the DirectX9 Alt blitter in the x64 version as I mentioned in my original post? Or maybe this is a known limitation of the x64 vit version?

Thank you so much again for your continuing work.

**Comment by dinkc64** [comment]
The directx9 alt blitter works fine in the 64bit version on my computer, it's probably down to the blitter requesting hw access to something that your video drivers don't have in 64bit mode.  I'm not really good with the blitter code, so it's only a guess.  Personally, I recommend using the 32bit builds regardless of system type.

best regards,
- dink

**Comment by dezraj** [comment]
Interesting. Here it works in the 32bit version, but not in 64 one. I'm pretty sure I have the June 2010 DirectX runtimes installed.

I have an Nvidia graphics card, a GTX 1660 Ti with driver version 516.94.

Do you think I should open a separate issue for that and close this one?

Anyway, for the time being I'll use the 32 bit version because it lets me use the DirectX 9 Alt blitter which has some filters I may play with eventually.

Thank you again for the fast reply.

**Comment by dinkc64** [comment]
Hi,
Regarding the DX9 Alt Blitter on the 64bit exe, could you get the Debug version, then try to switch to the alt blitter, when the error comes up, click OK then exit fbneo and attach the file "zzburndebug.html" here.  I did a little thinking about the problem, and came up with an idea to help figure out what is going wrong.  Maybe this will lead to some good ideas, let's see? :)

Idea #2:
after switching to alt blitter and clicking OK on the error:
see if you can still change some of the blitter options, for example, make sure HardFX is set to None, and SoftFX is double pixels (the first option), maybe a certain option is keeping it from working?

best regards,
- dink

**Comment by dezraj** [comment]
Hello @dinkc64 ! Thank you for your reply.

Is seems the debug version available in GitHub is only for the 32 bit version of the program, no? Does it work to try the 64 bit version too?

I ask because the issue I am having with the Direct9X Alt blitter only affects the x64 version, not the x86 one.

Thank you again for your time.

**Comment by barbudreadmon** [comment]
"github actions" failed to build the x64 debug for some reason (looks like a temporary failure), i'm currently re-running the failed job, let's hope it works this time.

**Comment by dinkc64** [comment]
Github is having some technical problems the past 2 days, hopefully they'll have things sorted soon!

**Comment by dezraj** [comment]
Thank you @dinkc64 . I could now download the x64 debug version.

I suppose I wasn't precise in my bug report. Apparently I can select the DirectX9 Alt blitter in the x64 version, but what I can't is select any of the SoftFX algorithms, except Double pixels (3D hardware).

Here's the log file when the error message showed and I quit the program:
[zzBurnDebug.zip](https://github.com/finalburnneo/FBNeo/files/11466574/zzBurnDebug.zip)

Actually I can select the SoftFX algorithm, but as soon as I start the game, it shows the error message and then the screen is black and the game is paused.

Let me know if I can help you with any other testing.

Thank you again for your time.

**Comment by dinkc64** [comment]
dezraz, thanks for the info's and log, I managed to get some of the softfx algo's working with the 64bit build, except scale2x.  I'll try to get that one working later today. :)

best regards,
- dink

**Comment by dinkc64** [comment]
p.s. the algo's in the list that are shaded are 16bit, and in the dx9 blitter, it does everything in 32bit, so 16bit scalers just won't work.

**Comment by dinkc64** [comment]
update: got a few more working on the 64bit exe :)  Please give it a try

best regards,
- dink

**Comment by dezraj** [comment]
It worked! Thank you for that!

It seems that the issues with the x64 build are gone then, except that the CRT 2x2 (Fast 480p) algorithm doesn't work in the DirectX9 Alt blitter, only in the SoftFX one. But this also happens in the 32x build, so perhaps it's a limitation of this blitter? This is what it looks like if I try to use this video effect:
![image](https://github.com/finalburnneo/FBNeo/assets/27839354/13169524-01f7-4a57-85f8-4725e62c5198)

And the DDT3x, on the other hand, doesn't work in the SoftFX blitter, only in the DirectX9 Alt one. It looks like this:
![image](https://github.com/finalburnneo/FBNeo/assets/27839354/b47f2878-8692-4227-bd0e-c915b4b360b9)

To summarize, the CRT 2x2 (Fast 480p) only works in SoftFX and the DDT3x only works in DirectX9 Alt (both 32 and 64 bit versions of the program).

Maybe this is already a known issue, but I'm reporting it anyway in case you want to leave a comment about it.

If you think everything is resolved you can close this issue. Thank you so much again.

**Comment by dezraj** [comment]
@taoenwen and @dinkc64 , I just realized that the KOF96 patch file "8-time.dat" is not working in FBNeo, but it does work in Raine. The time was supposed to pause in the character select screen and team order screen, but this is not happening.

In the file selector it's this one:
![image](https://github.com/finalburnneo/FBNeo/assets/27839354/1eaa0a12-cfd7-4520-8968-1d343f8e47fa)

You can get this patch in the YouTube video description in my original post.

Thank you so much for any help.

**Comment by dezraj** [comment]
Thank you @dinkc64 . The IPS patch to stop the time count in KOF96 is now working as well as the CRT2x2 (Fast 480p) in the DirectX9 Alt blitter. Only the DDT3x filter in SoftFX remains broken, but that's a very minor issue I guess.

You can close this issue whenever you think all problems reported here are fixed.

Thanks a lot again for your time and work.

**Comment by khanhnguyen1988** [comment]
Hello . fbneo file .dat can work with patch .bps , .ups...? I find it quite common of NES games

**Comment by taoenwen** [comment]
> Hello . fbneo file .dat can work with patch .bps , .ups...? I find it quite common of NES games

Can't, and there's no need to implement such a feature, no nes game is larger than 16MB

**Comment by khanhnguyen1988** [comment]
@taoenwen Hello . Recent emulator update makes some Streets of Rage 2 (Sega) game hacks unable to read through ips patch

**Comment by barbudreadmon** [comment]
@khanhnguyen1988 I haven't seen taoenwen for some time so i'm not entirely sure he'll answer, i hope nothing bad happened.

Is that issue specific to that ips patch or is it something affecting all ips hacks ?

**Comment by dinkc64** [comment]
@khanhnguyen1988 Hi, please post the hacks for Streets of Rage 2, I will try to fix it!

best regards,
- dink

**Comment by khanhnguyen1988** [comment]
@dinkc64 
sorry. it was my mistake. i edited the old .dat file to match the new roms info
streets of rage 2 (usa).bin old info
streets of rage 2 (USA)(1992)(Sega).bin new info
ips has been run on new emulator
![Screenshot 2024-10-31 223827](https://github.com/user-attachments/assets/609b05f5-3bf1-459f-b0d8-8cfc21e8faae)
![Screenshot 2024-10-31 224008](https://github.com/user-attachments/assets/e700cc9b-6b9a-4b59-b9cd-f47e2dec9569)

**Comment by ani917** [comment]
> [@taoenwen](https://github.com/taoenwen) Hello . Recent emulator update makes some Streets of Rage 2 (Sega) game hacks unable to read through ips patch

emulator reads a ROM with any name in the archive if the check matches, but then seems to rename it and patches only with a specific name, which in gameinfo

**Comment by ani917** [comment]
> > Hello . fbneo file .dat can work with patch .bps , .ups...? I find it quite common of NES games
> 
> Can't, and there's no need to implement such a feature, no nes game is larger than 16MB

game patches exist in this format for different games, such a feature would definitely be useful
